### PR TITLE
fix(Cardinality): Adding filodb partition check for cardinality queries 

### PR DESCRIPTION
**Pull request by sandeep6189**
**Pull Request checklist**

- [x] The commit(s) message(s) follows the contribution [guidelines](CONTRIBUTING.md) ?
- [x] Tests for the changes have been added (for bug fixes / features) ?
- [x] Docs have been added / updated (for bug fixes / features) ?

**Current behavior :**  We don't have any validation check for the cardinality queries. 

**New behavior :** 
1. Adding a filodb partition check for TsCardinalities queries.
2. Passing filodb-partition in QueryContext for Metering queries

### CODE REVIEW

File: core/src/main/scala/filodb.core/downsample/DownsampledTimeSeriesStore.scala [hunk 1]
```diff
@@ -73,8 +74,12 @@ extends TimeSeriesStore with StrictLogging {
     }
   }
 
-  def scanTsCardinalities(ref: DatasetRef, shards: Seq[Int],
+  def scanTsCardinalities(queryContext: QueryContext, ref: DatasetRef, shards: Seq[Int],
                           shardKeyPrefix: Seq[String], depth: Int): Seq[CardinalityRecord] = {
+    // adding an additional check to verify if the partition is same as mentioned in query context
+    require(isCorrectPartitionForCardinalityQuery(queryContext, FILODB_PARTITION),
```
Comment (yu-shipit) [inline core/src/main/scala/filodb.core/downsample/DownsampledTimeSeriesStore.scala:80]: It seems `queryContext` is only used for this newly added check.
Should this check happen before this function?

File: core/src/main/scala/filodb.core/store/ChunkSource.scala [hunk 2]
```diff
@@ -198,7 +198,26 @@ trait ChunkSource extends RawChunkSource with StrictLogging {
...
+
+  /**
+   * Additional check to ensure the query is for the given filodb partition
+   *
+   * @param queryContext QueryContext containing traceInfo of the originating partition
+   * @return
+   */
+  def isCorrectPartitionForCardinalityQuery(queryContext: QueryContext, filodbPartition: String): Boolean = {
+    if (
+      queryContext.traceInfo.contains(FILODB_PARTITION_KEY) &&
```
Comment (yu-shipit) [inline core/src/main/scala/filodb.core/store/ChunkSource.scala:211]: So the only invalid case is that we find a `trace info` does not match the `filodbPartition` from the `partition key`?

File: core/src/main/scala/filodb.core/downsample/DownsampledTimeSeriesStore.scala (hunk 1 above)
Comment (sandeep6189) [reply core/src/main/scala/filodb.core/downsample/DownsampledTimeSeriesStore.scala:80]: Yes. As of now, query context is not being used in TsCardinalities.  Thought of multiple ways, but querycontext was easier to pass without impacting the existing code flow.

We need this check when it is being executed on the filodb ( and also the component should have access to the config ). Before this, we are still in logical / exec plan realm.

File: core/src/main/scala/filodb.core/store/ChunkSource.scala (hunk 2 above)
Comment (sandeep6189) [reply core/src/main/scala/filodb.core/store/ChunkSource.scala:211]: Yes, that is the only invalid case. This is done to make sure the existing api flows in which TsCardinalities is queried is not impacted.

**Review by yu-shipit — APPROVED** [review]

**Pull request merged**